query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                // If the value is a variable or string, we must use `all_datoms`, or do the join
+                // ourselves, because we'll need to either extract or compare on the string.
+                Some(
+                    match value {
+                        // TODO: see if the variable is projected, aggregated, or compared elsewhere in
+                        // the query. If it's not, we don't need to use all_datoms here.
+                        &PatternValuePlace::Variable(_) =>
+                            DatomsTable::AllDatoms,       // TODO: check types.
+                        &PatternValuePlace::Constant(NonIntegerConstant::Text(_)) =>
+                            DatomsTable::AllDatoms,
+                        _ =>
+                            DatomsTable::Datoms,
+                    }),
+            &PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id)
+                      .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+            &PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw)
+                    .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+        }
+    }
+
+    /// Produce a (table, alias) pair to handle the provided pattern.
+    /// This is a mutating method because it mutates the aliaser function!
+    /// Note that if this function decides that a pattern cannot match, it will flip
+    /// `is_known_empty`.
+    fn alias_table<'s, 'a>(&mut self, schema: &'s Schema, pattern: &'a Pattern) -> Option<SourceAlias> {
+        self.table_for_places(schema, &pattern.attribute, &pattern.value)
+            .map(|table| SourceAlias(table, (self.aliaser)(table)))
+    }
+
+    fn get_attribute<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<&'s Attribute> {
+        match pattern.attribute {
+            PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id),
+            PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw),
+            _ =>
+                None,
+        }
+    }
+
+    fn get_value_type<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<ValueType> {
+        self.get_attribute(schema, pattern).map(|x| x.value_type)
+    }
+}
+
+/// Expansions.
+impl ConjoiningClauses {
+
+    /// Take the contents of `bindings` and generate inter-constraints for the appropriate
+    /// columns into `wheres`.
+    /// For example, a bindings map like
+    ///
+    /// ```clojure
+    ///   {?foo [:datoms12.e :datoms13.v :datoms14.e]}
+    /// ```
+    ///
+    /// produces additional constraints:
+    ///
+    /// ```clojure
+    /// [[:= :datoms12.e :datoms13.v] [:= :datoms12.e :datoms14.e]]
+    /// ```
+    pub fn expand_bindings(&mut self) {
+        for cols in self.bindings.values() {
+            if cols.len() > 1 {
+                let ref primary = cols[0];
```

Comment:
This is not an ambiguous case, as far as I can tell -- and the discussion you link to seems to conclude that the two are equivalent: https://github.com/rust-lang/rust-by-example/issues/390#issuecomment-129167712.